pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -390,245 +427,233 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, MultiRewa
     }
 
-    function _updateDistribution(IERC20 stakingToken, address user) internal {
+    function _updateDistributions(IERC20 stakingToken, address user) internal {
         UserStaking storage userStaking = _userStakings[stakingToken][user];
-        EnumerableSet.Bytes32Set storage distributions = userStaking.allowedDistributions;
+        EnumerableSet.Bytes32Set storage distributions = userStaking.subscribedDistributions;
         uint256 distributionsLength = distributions.length();
 
         for (uint256 i; i < distributionsLength; i++) {
             bytes32 distributionId = distributions.unchecked_at(i);
-            _updateUserRewardRatePerToken(userStaking, distributionId);
+            _updateUserTokensPerStake(userStaking, distributionId);
         }
     }
 
-    /* ========== MODIFIERS ========
+    function _updateUserTokensPerStake(UserStaking storage userStaking, bytes32 distributionId) internal {
+        uint256 globalTokensPerStake = _updateDistributionRate(distributionId);
+        UserDistribution storage userDistribution = userStaking.distributions[distributionId];
+        userDistribution.unclaimedTokens = _totalUnclaimedTokens(userStaking, distributionId);
+        userDistribution.userTokensPerStake = globalTokensPerStake;
+    }
 
     /**
-     * @notice Updates the reward rate for all the distributions a user has signed up for
+     * @notice Updates the amount of distribution tokens paid per token staked for a distribution
+     * @dev This is expected to be called whenever a user's applicable staked balance changes,
+     *      either through adding/removing tokens or subscribing/unsubscribing from the distribution.
+     * @param distributionId ID of the distribution being updated
+     * @return globalTokensPerStake The updated number of distribution tokens paid per staked token
      */
-    modifier updateDistribution(IERC20 stakingToken, address user) {
-        _updateDistribution(stakingToken, user);
-        _;
+    function _updateDistributionRate(bytes32 distributionId) internal returns (uint256 globalTokensPerStake) {
+        Distribution storage distribution = _getDistribution(distributionId);
+        globalTokensPerStake = _globalTokensPerStake(distribution);
+        distribution.globalTokensPerStake = globalTokensPerStake;
+        distribution.lastUpdateTime = _lastTimePaymentApplicable(distribution);
+    }
+
+    function _globalTokensPerStake(Distribution storage distribution) internal view returns (uint256) {
+        uint256 supply = distribution.totalSupply;
+        if (supply == 0) {
+            return distribution.globalTokensPerStake;
+        }
+
+        // Underflow is impossible here because _lastTimePaymentApplicable(...) is always greater than last update time
+        uint256 unpaidDuration = _lastTimePaymentApplicable(distribution) - distribution.lastUpdateTime;
+        uint256 unpaidAmountPerToken = Math.mul(unpaidDuration, distribution.paymentRate).divDown(supply);
```

Comment:
I found the fixed point division here bizarre. Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/1010 to track this effort.

paymentRate is created as total / duration (non-fp). So what this does is:

`(elapsed * total / duration) * 1e18 / supply` (the 1e18 comes from fixed point division)

which seems incorrect?

Reply:
Yes, I saw this and thought it was weird as it's the only usage of `FixedPoint` in the contract. Definitely going to dig into this more but I didn't want to muck with maths in this PR.

Reply:
Agreed, we'll follow up in that issue!